### PR TITLE
Incorporate LCache library as the persistent storage backend

### PR DESCRIPTION
**Pull request by danielbachhuber**

### CODE REVIEW

File: object-cache.php [hunk 1]
```diff
@@ -876,15 +908,48 @@ public function wp_action_admin_notices_warn_missing_apcu() {
...
+				}
+			}
+
+			if ( defined( 'WP_LCACHE_RUNNING_TESTS' ) && WP_LCACHE_RUNNING_TESTS ) {
+				wp_lcache_initialize_database_schema();
+			}
+
+			$dsn = 'mysql:host='. DB_HOST. ';port='. $port .';dbname='. DB_NAME;
+			$options = array( PDO::ATTR_TIMEOUT => 2, PDO::MYSQL_ATTR_INIT_COMMAND => 'SET sql_mode="ANSI_QUOTES"' );
+			$dbh = new PDO( $dsn, DB_USER, DB_PASSWORD, $options );
```
Comment (danielbachhuber) [inline object-cache.php:955]: @davidstrauss Is there a particular reason we need to use PDO?

File: object-cache.php [hunk 2]
```diff
@@ -876,15 +908,48 @@ public function wp_action_admin_notices_warn_missing_apcu() {
...
+						$socket = substr( $maybe_socket, 1 );
+					}
+				} else {
+					$socket = $port_or_socket;
+				}
+			}
+
+			if ( defined( 'WP_LCACHE_RUNNING_TESTS' ) && WP_LCACHE_RUNNING_TESTS ) {
+				wp_lcache_initialize_database_schema();
+			}
```
Comment (danielbachhuber) [inline object-cache.php:951]: Unfortunately, this is the only way to create the database tables at the appropriate time in the test suite. There's no hook we can use between database initialization and object cache initialization.

**Comment by danielbachhuber** [comment]
@stevector This is ready for an initial review.

File: object-cache.php [hunk 3]
```diff
@@ -822,52 +821,85 @@ protected function _should_persist( $group ) {
...
-	 * Admin UI to let the end user know that APCu isn't available
+	 * Admin UI to let the end user know that LCache isn't available
 	 */
-	public function wp_action_admin_notices_warn_missing_apcu() {
+	public function wp_action_admin_notices_warn_missing_lcache() {
 		if ( ! current_user_can( 'manage_options' ) ) {
 			return;
 		}
-		echo '<div class="message error"><p>Warning! APCu module is unavailable, which is required by WP LCache object cache.</p></div>';
+		echo '<div class="message error"><p>Warning! LCache library or APCu is unavailable, which is required by WP LCache object cache.</p></div>';
```
Comment (danielbachhuber) [inline object-cache.php:899]: We should provide a message descriptive of the specific error.

**Comment by stevector** [comment]
Thanks @danielbachhuber. I may not get to this PR until late next week or the week after. I'm juggling a couple of other tasks. I saw you chatting with @joshkoenig about this PR too. He might have feedback before I do. I'm curious for @ataylorme's thoughts too.

File: object-cache.php (hunk 1 above)
Comment (joshkoenig) [inline object-cache.php:955]: It should be fine to use the WP query abstraction.

**Comment by joshkoenig** [comment]
Looks like test failures with the delete command.

Yeah in the core lib we have it calling `l1->delete()` on line 566 of `lcache.php`. But the `delete()` function signature currently requires an `$event_id`.

**Comment by danielbachhuber** [comment]
> Yeah in the core lib we have it calling `l1->delete()` on line 566 of `lcache.php`. But the `delete()` function signature currently requires an `$event_id`.

Pushed a fix to make `$event_id` optional, which isn't used in that method.

File: object-cache.php [hunk 4]
```diff
@@ -525,10 +528,10 @@ public function get( $key, $group = 'default', $force = false, &$found = null )
 		}
 
 		$id = $this->_key( $key, $group );
-		$value = $this->_call_apcu( 'apcu_fetch', $id );
+		$value = $this->_call_lcache( 'get', $id );
 
-		// APCu returns `false` when the key doesn't exist
-		if ( false === $value ) {
+		// LCache returns `false` when the key doesn't exist
```
Comment (stevector) [inline object-cache.php]: @danielbachhuber this comment should say "null", right?

Comment (danielbachhuber) [inline object-cache.php]: > this comment should say "null", right?

Correct. Updated 0d54520

File: object-cache.php [hunk 5]
```diff
@@ -876,15 +914,52 @@ public function wp_action_admin_notices_warn_missing_apcu() {
...
 		$this->multisite = is_multisite();
 		$this->blog_prefix = $this->multisite ? $blog_id . ':' : '';
 
-		$this->is_apcu_available = function_exists( 'apcu_sma_info' ) && apcu_sma_info();
+		// apcu_sma_info() triggers a warning when APCu is disabled
+		// @codingStandardsIgnoreStart
+		if ( function_exists( 'apcu_sma_info' ) && @apcu_sma_info() && class_exists( 'LCacheNullL1' ) ) {
+		// @codingStandardsIgnoreEnd
+			$l1 = new LCacheNullL1();
+			if ( php_sapi_name() !== 'cli' || 'on' === ini_get( 'apc.enable_cli' ) ) {
```
Comment (stevector) [inline object-cache.php:928]: Why is the CLI relevant here? Can you add a comment?

Comment (danielbachhuber) [inline object-cache.php:928]: > Why is the CLI relevant here?

APCu isn't available in CLI context unless explicitly enabled

> Can you add a comment?

Yep a6cccf7

File: object-cache.php [hunk 6]
```diff
@@ -876,15 +914,52 @@ public function wp_action_admin_notices_warn_missing_apcu() {
...
+				if ( 0 !== strpos( $port_or_socket, '/' ) ) {
+					$port = intval( $port_or_socket );
+					$maybe_socket = strstr( $port_or_socket, ':' );
+					if ( ! empty( $maybe_socket ) ) {
+						$socket = substr( $maybe_socket, 1 );
+					}
+				} else {
+					$socket = $port_or_socket;
+				}
+			}
```
Comment (danielbachhuber) [inline object-cache.php:947]: I stole this out of `class wpdb`, but we might just want to use `$wpdb` instead.

**Pull request merged**